adr/ADR-0003-trustyai-service-deployment-using-operator-pattern.md

```diff
@@ -0,0 +1,245 @@
+
+In addition to this, the deployment and the storage (PVC for now) must be created into a user owned namespace to give users full control and prevent security issues. An operator can enforce this.
+
+## Goals
+
+* Automate the deployment, management, and maintenance of the TrustyAI service.
+* Reduce manual errors and increase consistency in deployments.
+* Help updating the TrustyAI service.
+
+## Non-goals
+
+Implementing mechanisms that perform actions unrelated with the lifecycle of the TrustyAI service (create, upgrade, monitor, etc)..
+
+## Current situation
+
+Currently, TrustyAI service deployments are done manually or through scripts that do not fully take advantage of Kubernetes. This can be inefficient and lead to errors introduced by manual steps.
+
+As an example, the TrustyAI service needs to update ModelMesh's configuration to add the TrustyAI service as a new endpoint. This is currently done via a deployment-time script that patches the ModelMesh configuration. This could be done automatically by the Operator.
+
+Although TrustyAI's deployment needs to configure a considerable number of resources (_e.g._ `Deployment`, `Service`, `ConfigMap`, `Route`, `ServiceMonitor`), the actual configuration options available for a custom TrustyAI deployment are limited. This means that a custom TrustyAI Custom Resource Definition (CRD) would be quite simple, and the Operator would be able to handle the creation and management of the required resources.
+
+## Proposal
+
+We propose to use a stand-alone TrustyAI Kubernetes Operator which would create and manage the required Deployment, Service, ConfigMap, Route, and ServiceMonitor resources based on a simple Custom Resource while keeping the state consistent with the desired one [^1].
+
+[^1]: Initial implementation at https://github.com/ruivieira/trustyai-service-operator
+
+### Custom Resource
+
+An example of a custom resource is:
+
+```yaml
+apiVersion: trustyai.opendatahub.io/v1alpha1
+kind: TrustyAIService
+metadata:
+  name: trustyai-service-example
+  
+spec:
+  storage:
+    format: "PVC"
+    folder: "/inputs"
+  data:
+    filename: "data.csv"
+    format: "CSV"
+  trustyaiMetrics:
+    schedule: "15s"
+status:
+  phase: …
+  replicas: …
+  conditions:
+  - type: Ready
+    …
+  - type: ModelMeshReady 
+    status: "True"
+    lastTransitionTime: …
+    reason: ModelMeshHealthy
+    message: ModelMesh is running and healthy.
+  - type: StorageReady
+    status: "True"
+    lastTransitionTime: …
+    reason: StorageHealthy
+    message: Storage system is functioning correctly.
+    lastUpdateTime: …
+```
+
+In this example:
+
+* `replicas` is an optional field that specifies the number of replicas of the TrustyAI service that you want to run. If not provided, the default is one replica.
+* `storage` is a mandatory field that specifies the storage details. It has two nested fields:
+  * `format` - the storage format, (example: a Persistent Volume Claim (PVC)).
```

Comment:
Is there a specific AccessMode PVC needed? (RWX,RWO)

Reply:
- Is it using default StorageClass? 
- So default StorageClass will be a prerequisite?
- If there is no default storageClass, what will happen?

Reply:
@Jooho Good point, I'll add this info.
RWO is what the manifests have been specifying so far, so I think we could keep with that.
Regarding the StorageClass, the initial implementation disables dynamic provisioning and binds to already existing PVs.
Along the lines of

```
spec:
  storage:
    pv: "mypv"
    size: 1Gi
```